src/main/java/com/minecolonies/coremod/colony/buildings/inventory/ItemStackHandlerWithIndex.java

```diff
@@ -0,0 +1,333 @@
+package com.minecolonies.coremod.colony.buildings.inventory;
+
+import com.minecolonies.api.crafting.ItemStorage;
+import com.minecolonies.api.util.ItemStackUtils;
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.NonNullList;
+import net.minecraftforge.items.ItemStackHandler;
+import org.jetbrains.annotations.NotNull;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Set;
+import java.util.function.Predicate;
+
+/**
+ * ItemStackHandler which also keeps an index of its contents.
+ * It also can push any changes to a meta(upper) index, to combine multiple handlers.
+ *
+ * @param <Value> The type of the meta index, not required.
+ */
+public class ItemStackHandlerWithIndex<Value> extends ItemStackHandler
```

Comment:
Any public member should be pulled up into an interface that extends ItemStackHandler.

Reply:
There are too many issues with this class to count.......
The implementation seems to depend on meta data leaving NBT out of side entirely, the indexing is weirdly implemented, and as far as I can tell it is not compatible with any existing IItemHandler implementation, which is what is actually needed.

In my opinion it would be much better if we created a global index off all items in a colony in every IItemHandler.
Update one inventory per tick (configurable) and use that to perform lookups....

That allows us to incorporate inventories from other mods as well, without having to screw around.

And it is the same logic that mods like AE or RefStorage apply.

Reply:
meta might've been bad wording, but it isnt actually related to itemstacks/NBT but instead is a reference to a global index which is updated too when this index gets updated. So you get local index for each rack and a global for each building which are kept in sync

Reply:
The first of all your naming is flaw as hell.
Second of all why the two tier system?